File: Assets/Plugins/Android/PermissionPlease/Java/PermissionPlease.java
/*
 * Permission Please v0.9
 * By: Jafar Abdulrasoul [Jimmar]
 * Original code from https://stackoverflow.com/questions/35027043/implementing-android-6-0-permissions-in-unity3d
 * Special thanks to Noodlecake studios
 */

package net.jimmar.unityplugins;

import android.Manifest;
import android.app.Activity;
import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;


import com.unity3d.player.UnityPlayer;

public class PermissionPlease
{
    private final static String UNITY_CALLBACK_GAMEOBJECT_NAME = "PermissionPlease";
    private final static String UNITY_CALLBACK_METHOD_NAME = "PermissionRequestCallbackInternal";
    public final static String PERMISSION_GRANTED = "PERMISSION_GRANTED";
    public final static String PERMISSION_DENIED = "PERMISSION_DENIED";
    private final static String[] permissionsArr = {
            //Calendar
            Manifest.permission.READ_CALENDAR,
            Manifest.permission.WRITE_CALENDAR,
            //Camera
            Manifest.permission.CAMERA,
            //Contacts
            Manifest.permission.READ_CONTACTS,
            Manifest.permission.WRITE_CONTACTS,
            Manifest.permission.GET_ACCOUNTS,
            //Location
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            //Microphone
            Manifest.permission.RECORD_AUDIO,
            //Phone
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.READ_PHONE_NUMBERS,
            Manifest.permission.CALL_PHONE,
            Manifest.permission.ANSWER_PHONE_CALLS,
            Manifest.permission.READ_CALL_LOG,
            Manifest.permission.WRITE_CALL_LOG,
            Manifest.permission.ADD_VOICEMAIL,
            Manifest.permission.USE_SIP,
            Manifest.permission.PROCESS_OUTGOING_CALLS,
            //Sensors
            Manifest.permission.BODY_SENSORS,
            //SMS
            Manifest.permission.SEND_SMS,
            Manifest.permission.RECEIVE_SMS,
            Manifest.permission.READ_SMS,
            Manifest.permission.RECEIVE_WAP_PUSH,
            Manifest.permission.RECEIVE_MMS,
            //Storage
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE,
    };
    
    
    public static boolean logging = false;
    
    public static String getPermissionStringFromEnumInt(int permissionEnum) throws Exception
    {
        if(permissionEnum < permissionsArr.length)
            return permissionsArr[permissionEnum];

        if(logging) Log.e("[PermissionPlease]", "Error. Unknown permissionEnum " + permissionEnum);
        throw new Exception(String.format("Error. Unknown permissionEnum %d",permissionEnum));
    }
    
    public static void UnitySendMessage(String message){
        UnityPlayer.UnitySendMessage(UNITY_CALLBACK_GAMEOBJECT_NAME, UNITY_CALLBACK_METHOD_NAME, message);
    }
    
    public static void grantPermission(Activity currentActivity, int permissionEnum, boolean enableLogging)
    {
        logging = enableLogging;
        if(logging) Log.i("[PermissionPlease]","grantPermission " + permissionEnum);
        if (Build.VERSION.SDK_INT < 23) {
            if(logging) Log.i("[PermissionPlease]","Build.VERSION.SDK_INT < 23 (" + Build.VERSION.SDK_INT+")");
            UnitySendMessage(PERMISSION_GRANTED);
            return;
        }

        try
        {
            final int PERMISSIONS_REQUEST_CODE = permissionEnum;
            final String permissionFromEnumInt = getPermissionStringFromEnumInt(permissionEnum);
            if (currentActivity.checkCallingOrSelfPermission(permissionFromEnumInt) == PackageManager.PERMISSION_GRANTED) {
                if(logging) Log.i("[PermissionPlease]", "already granted");
                UnitySendMessage(PERMISSION_GRANTED);
                return;
            }

            final FragmentManager fragmentManager = currentActivity.getFragmentManager();
            final Fragment request = new InnerFragment(permissionEnum, permissionFromEnumInt, fragmentManager);

            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.add(0, request);
            fragmentTransaction.commit();
        }
        catch(Exception error)
        {
            if(logging) Log.w("[PermissionPlease]", String.format("Unable to request permission: %s", error.getMessage()));
            UnitySendMessage(PERMISSION_DENIED);
        }
    }

    public static class InnerFragment extends Fragment{
        String permissionFromEnumInt;
        int PERMISSIONS_REQUEST_CODE;
        FragmentManager fragmentManager;

        public InnerFragment(int permissionEnum, String permissionFromEnumInt, FragmentManager fragmentManager){
            PERMISSIONS_REQUEST_CODE = permissionEnum;
            this.permissionFromEnumInt = permissionFromEnumInt;
            this.fragmentManager = fragmentManager;
        }

        @Override public void onStart()
        {
            super.onStart();
            String[] permissionsToRequest = new String [] {permissionFromEnumInt};
            requestPermissions(permissionsToRequest, PERMISSIONS_REQUEST_CODE);
        }

        @Override public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults)
        {
            if (requestCode != PERMISSIONS_REQUEST_CODE)
                return;

            if (grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // permission granted
                if(logging) Log.i("[PermissionPlease]", PERMISSION_GRANTED);
                UnitySendMessage(PERMISSION_GRANTED);
            } else {

                // permission denied
                if(logging) Log.i("[PermissionPlease]",PERMISSION_DENIED);
                UnitySendMessage(PERMISSION_DENIED);
            }

            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.remove(this);
            fragmentTransaction.commit();
        }
    }
}

File: Assets/Plugins/Android/PermissionPlease.cs
/*
 * Permission Please v0.9
 * By: Jafar Abdulrasoul [Jimmar]
 * Original code from https://stackoverflow.com/questions/35027043/implementing-android-6-0-permissions-in-unity3d
 * Special thanks to Noodlecake studios
 */

using System;
using UnityEngine;

namespace Plugins.Android.PermissionPlease
{
    public class PermissionPlease : MonoBehaviour
    {
        private static Action<bool> PermissionRequestCallback;

        public enum AndroidPermission
        {
            //Calendar
            READ_CALENDAR,
            WRITE_CALENDAR,
            //Camera
            CAMERA,
            //Contacts
            READ_CONTACTS,
            WRITE_CONTACTS,
            GET_ACCOUNTS,
            //Location
            ACCESS_FINE_LOCATION,
            ACCESS_COARSE_LOCATION,
            //Microphone
            RECORD_AUDIO,
            //Phone
            READ_PHONE_STATE,
            READ_PHONE_NUMBERS,
            CALL_PHONE,
            ANSWER_PHONE_CALLS,
            READ_CALL_LOG,
            WRITE_CALL_LOG,
            ADD_VOICEMAIL,
            USE_SIP,
            PROCESS_OUTGOING_CALLS,
            //Sensors
            BODY_SENSORS,
            //SMS
            SEND_SMS,
            RECEIVE_SMS,
            READ_SMS,
            RECEIVE_WAP_PUSH,
            RECEIVE_MMS,
            //Storage
            WRITE_EXTERNAL_STORAGE,
            READ_EXTERNAL_STORAGE,
        }

        public static void GrantPermission(AndroidPermission permission, Action<bool> requestCallback = null,
            bool enableLogging = false)
        {
            #if !UNITY_ANDROID
                Debug.LogWarning("Only Android platform is supported");
                return;
            #endif
            
            #if UNITY_EDITOR
                Debug.LogWarning("Editor is not supported, test on an actual device or an Android Emulator");
                return;
            #endif
            
            if (!initialized)
                initialize();
            PermissionRequestCallback = requestCallback;
            PermissionPleaseClass.CallStatic("grantPermission", activity, (int) permission, enableLogging);
        }

        //////////////////////////////
        /// Initialization Stuff /////
        //////////////////////////////

        // it's a singleton, but no one needs to know about it. hush hush. dont touch me.
        private static PermissionPlease instance;

        private static bool initialized;

        public void Awake()
        {
            // instance is also set in initialize.
            // this ensures that it doesn't break if the component was added to the scene manually.
            instance = this;
            DontDestroyOnLoad(gameObject);
            // object name must match UnitySendMessage call in PermissionPlease.java
            if (name != PERMISSION_PLEASE_NAME)
                name = PERMISSION_PLEASE_NAME;
        }

        private static void initialize()
        {
            // runs once when you call GrantPermission
            if (instance == null)
            {
                GameObject go = new GameObject();
                // instance will also be set in awake, but having it here as well seems extra safe
                instance = go.AddComponent<PermissionPlease>();
                // object name must match UnitySendMessage call in PermissionPlease.java
                go.name = PERMISSION_PLEASE_NAME;
            }

            // get the jni stuff. we need the activty class and the PermissionPlease class.
            PermissionPleaseClass = new AndroidJavaClass("net.jimmar.unityplugins.PermissionPlease");
            AndroidJavaClass u3d = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
            activity = u3d.GetStatic<AndroidJavaObject>("currentActivity");

            initialized = true;
        }


        ///////////////////
        //// JNI Stuff ////
        ///////////////////

        private static AndroidJavaClass PermissionPleaseClass;
        private static AndroidJavaObject activity;
        private const string PERMISSION_GRANTED = "PERMISSION_GRANTED"; // must match PermissionPlease.java
        private const string PERMISSION_DENIED = "PERMISSION_DENIED"; // must match PermissionPlease.java

        // must match UnitySendMessage call in PermissionPlease.java
        private const string PERMISSION_PLEASE_NAME = "PermissionPlease";

        private void PermissionRequestCallbackInternal(string message)
        {
            // this method is caleld from the java side.
            // the method name and gameobject name must match PermissionPlease.java's UnitySendMessage
            bool permissionGranted = (message == PERMISSION_GRANTED);
            if (PermissionRequestCallback != null)
                PermissionRequestCallback(permissionGranted);
        }
    }
}